### ISSUE TITLE
Update types over at @types/emoji-mart for v5

### ISSUE THREAD

**Issue by Scalahansolo** [missive/emoji-mart#576]
I might be able to find some time to work through this. But... I figured a maintainer would be able to blow this considerably faster.

More than anything, I wanted to a tracking ticket open in this repo for tracking purposes.

https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/emoji-mart

**Comment by EtienneLem** [comment]
Indeed! It would most likely be more convenient if we were to include types right here, right? It would be easier to maintain at the source I assume.

**Comment by Scalahansolo** [comment]
It would definitely be nice to ship releases of EmojiMart 5+ with a `.d.ts` file along with the release as to not have to install the disconnected dep from @types/emoji-mart

**Comment by Scalahansolo** [comment]
Another option here would be to just convert the codebase to TypeScript.... If that's something y'all would be open too, I could make do that work. But would want to make sure the Missive crew would be okay with that before I put in the time.

**Comment by EtienneLem** [comment]
Our Parcel scripts have the `--no-autoinstall` flag, you most likely need to add the `@parcel/transformer-typescript-types` to `packages/emoji-mart/package.json`

**Comment by Scalahansolo** [comment]
![ScreenShot 04-27 at 11 52](https://user-images.githubusercontent.com/4317253/165589061-4a025c24-a280-46b8-ab73-c127db3d7cbc.png)

I gave that a shot and still running into the same issue.

Also tried a diff plugin based on their docs in a .parcelrc, no luck there either 

![ScreenShot 04-27 at 11 53](https://user-images.githubusercontent.com/4317253/165589186-14a4952d-3947-4b86-b567-754d228d2973.png)

**Comment by EtienneLem** [comment]
Yeah I see, I’m getting the same error on `yarn build`. There must be something specific about `types` in Parcel codebase. From what I can see the `@parcel/transformer-typescript-types` does get properly installed in `root/node_modules` (as is expected)

**Comment by Scalahansolo** [comment]
On my local branch, I actually made some good progress on converting the whole lib to typescript. If we could get past this build issue / get y'alls blessing on move the project to typescript, I would be happy to finish that work out.

**Comment by zhouzi** [comment]
@Scalahansolo could you open a pull request with your progress? I am looking forward to the types being added and having your code would make it easier to help with the build issue 🙏

**Comment by Scalahansolo** [comment]
Yup. I just pushed a lingering commit I had locally. There is still a good amount of stuff to fix for converting the codebase to Typescript. I just haven't gone any further as I haven't heard back from maintainers if moving to typescript would be something they want.

**Comment by zhouzi** [comment]
I was able to get past the issue with parcel and opened #600 as a starting point.

**Comment by EtienneLem** [comment]
> I just haven't gone any further as I haven't heard back from maintainers if moving to typescript would be something they want.

Sorry about that. Migrating the codebase to typescript is something I would support.

**Comment by xmsz** [comment]
any update?

**Comment by sayjeyhi** [comment]
I didn't migrate to v5 because of this...

**Comment by rrfaria** [comment]
I have same issue, I'm using v5 with react and typescript.

Component works but to build it complains about types 

```
Could not find a declaration file for module '@emoji-mart/react'
/@emoji-mart+react@1.0.0_kbqdcrm7a2jamcyb7krk3b53dq/node_modules/@emoji-mart/react/dist/main.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/emoji-mart__react` if it exists or add a new declaration (.d.ts) file containing `declare module '@emoji-mart/react';`ts(7016)
```

**Comment by rrfaria** [comment]
I found a temporary solution 

install 
```
npm i --save-dev @types/emoji-mart
```

create a custom component

```ts
// EmojiPicker.tsx
import React, { useEffect, useRef } from 'react';

import data from '@emoji-mart/data';
import { Picker, PickerProps } from 'emoji-mart';
import i18n from '@emoji-mart/data/i18n/pt.json';
i18n.search_no_results_1 = 'Sem Resultados';

const EmojiPicker = (props: PickerProps | Readonly<PickerProps> | any) => {
  const ref = useRef() as React.MutableRefObject<HTMLInputElement>;

  useEffect(() => {
    new Picker({ ...props, data, ref, i18n });
  }, [props]);

  return <div ref={ref} />;
};

export default EmojiPicker;
```

then you can use it 
```
const yourFuncToHandleEmoji = (emoji: { native: string }) => {
    setText((state) => state + emoji.native);
  };

<EmojiPicker onEmojiSelect={yourFuncToHandleEmoji} />
```

problem solved

**Comment by justinbhopper** [comment]
@rrfaria Most have already been using `@types/emoji-mart` for v4 and older, but this ticket is about the fact that it is out of date since v5 and thus many of the interfaces (including `PickerProps`) are not up-to-date.  Your problem and solution are because you weren't using the type defs in the first place.

**Comment by gajus** [comment]
It would be easier to just convert this codebase to TypeScript than maintaining those separately.

**Comment by lgenzelis** [comment]
Same here. I'm still using v3 because that's the latest version which types (via `@types/emoji-mart`). Are there any plans regarding update the types?

**Comment by jwhitham-ds** [comment]
https://github.com/missive/emoji-mart/pull/671#issuecomment-1223342236

**Comment by zaadmin** [comment]
> 



> I have same issue, I'm using v5 with react and typescript.
> 
> Component works but to build it complains about types
> 
> ```
> Could not find a declaration file for module '@emoji-mart/react'
> /@emoji-mart+react@1.0.0_kbqdcrm7a2jamcyb7krk3b53dq/node_modules/@emoji-mart/react/dist/main.js' implicitly has an 'any' type.
>   Try `npm i --save-dev @types/emoji-mart__react` if it exists or add a new declaration (.d.ts) file containing `declare module '@emoji-mart/react';`ts(7016)
> ```

**Comment by huarmenta** [comment]
Would this be an acceptable way to bypass the missing types and just taking them from the class?

```ts
import type { Picker } from 'emoji-mart';

type ExtractProps<T> = {
  [K in keyof T]: T[K] extends { value: infer V } ? V : never;
};

type PickerPropsValues = ExtractProps<typeof Picker.Props>;
```

**Comment by ioma8** [comment]
How is this looking? I would really welcome the ts definitions have in the package. @types/emoji-mart doesnt work for me, probly cause it is related to older version or not to the react subpackage.

**Comment by rugk** [comment]
BTW at DT there is also the discussion here: https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/60094

I suggested to deprecate/remove the types there.

**Comment by rugk** [comment]
@huarmenta Cool idea, but for me it shows and cannot import the type:
```
Module '"emoji-mart"' has no exported member 'Picker'.
``` 

![Image](https://github.com/user-attachments/assets/d6061ccc-ebc6-4640-a292-0f36176f304a)

With `package.json`:
```json
  "dependencies": {
    "@emoji-mart/data": "^1.2.1",
    "emoji-mart": "^5.6.0"
  }
```

Maybe because only `emoji-mart/dist` is installed?

**Comment by rugk** [comment]
BTW found the TS definition for `PickerProps` in this repository: https://github.com/missive/emoji-mart/blob/16978d04a766eec6455e2e8bb21cd8dc0b3c7436/packages/emoji-mart/src/components/Picker/PickerProps.ts

**Comment by rugk** [comment]
Ah notably the _Props_ only apply to the WebComponent.

As for the headless search via `SearchIndex`, here the typings already exist: https://github.com/missive/emoji-mart/blob/main/packages/emoji-mart-data/index.d.ts (**edit:** or only for the data, there are some minor differences, see https://github.com/missive/emoji-mart/pull/996)
